File: lib/core/models/notification_preferences.dart
enum NotificationCategory {
  warAttacks,
  warState,
  warReminders,
  raidReminders,
  events,
  announcements,
  monthlySupport,
}

enum NotificationAccountSource {
  verified;

  static NotificationAccountSource fromWire(String value) => switch (value) {
    'verified' => NotificationAccountSource.verified,
    _ => throw FormatException('Unsupported notification account source'),
  };
}

class NotificationAccount {
  const NotificationAccount({
    required this.playerTag,
    required this.source,
    this.active = true,
  });

  final String playerTag;
  final NotificationAccountSource source;
  final bool active;

  factory NotificationAccount.fromJson(Map<String, dynamic> json) {
    final playerTag = json['playerTag'];
    final source = json['source'];
    final active = json['active'];
    if (playerTag is! String || playerTag.isEmpty || source is! String) {
      throw const FormatException('Invalid notification account');
    }
    return NotificationAccount(
      playerTag: playerTag,
      source: NotificationAccountSource.fromWire(source),
      active: active is bool ? active : true,
    );
  }
}

class NotificationPreferences {
  const NotificationPreferences({
    this.deviceId = '',
    this.environment = 'production',
    this.notificationsEnabled = false,
    this.warAttacks = false,
    this.warState = false,
    this.warReminders = false,
    this.raidReminders = false,
    this.events = false,
    this.announcements = false,
    this.monthlySupport = false,
    this.reminderTimings = const [],
    this.raidReminderTimings = const [],
    this.accounts = const [],
  });

  final String deviceId;
  final String environment;
  final bool notificationsEnabled;
  final bool warAttacks;
  final bool warState;
  final bool warReminders;
  final bool raidReminders;
  final bool events;
  final bool announcements;
  final bool monthlySupport;
  final List<int> reminderTimings;
  final List<int> raidReminderTimings;
  final List<NotificationAccount> accounts;

  factory NotificationPreferences.fromJson(Map<String, dynamic> json) {
    bool readBool(String key) {
      final value = json[key];
      if (value is! bool) {
        throw FormatException('Invalid $key');
      }
      return value;
    }

    final deviceId = json['deviceId'];
    final environment = json['environment'];
    final reminderTimings = json['reminderTimings'];
    final raidReminderTimings = json['raidReminderTimings'];
    final accounts = json['accounts'];
    if (deviceId is! String ||
        environment is! String ||
        reminderTimings is! List ||
        raidReminderTimings is! List ||
        accounts is! List) {
      throw const FormatException('Invalid notification preferences');
    }

    final minutes = reminderTimings
        .map((value) {
          if (value is! num || value.toInt() != value) {
            throw const FormatException('Invalid reminder timing');
          }
          return value.toInt();
        })
        .toList(growable: false);
    if (minutes.length > 3 ||
        minutes.any((value) => value < 1 || value > 2820) ||
        minutes.toSet().length != minutes.length) {
      throw const FormatException('Invalid reminder timings');
    }
    final parsedAccounts = accounts
        .map((value) {
          if (value is! Map) {
            throw const FormatException('Invalid notification account');
          }
          return NotificationAccount.fromJson(Map<String, dynamic>.from(value));
        })
        .toList(growable: false);
    final raidMinutes = raidReminderTimings
        .map((value) {
          if (value is! num || value.toInt() != value) {
            throw const FormatException('Invalid Raid Weekend reminder timing');
          }
          return value.toInt();
        })
        .toList(growable: false);
    if (raidMinutes.length > 3 ||
        raidMinutes.any(
          (value) => value < 1 || value > 4320 || value % 15 != 0,
        ) ||
        raidMinutes.toSet().length != raidMinutes.length) {
      throw const FormatException('Invalid Raid Weekend reminder timings');
    }

    return NotificationPreferences(
      deviceId: deviceId,
      environment: environment,
      notificationsEnabled: readBool('notificationsEnabled'),
      warAttacks: readBool('warAttacksEnabled'),
      warState: readBool('warStateEnabled'),
      warReminders: readBool('warRemindersEnabled'),
      raidReminders: readBool('raidRemindersEnabled'),
      events: readBool('eventsEnabled'),
      announcements: readBool('announcementsEnabled'),
      monthlySupport: readBool('monthlySupportEnabled'),
      reminderTimings: minutes,
      raidReminderTimings: raidMinutes,
      accounts: parsedAccounts,
    );
  }

  Map<String, dynamic> toPutJson({
    required String deviceId,
    required String environment,
  }) {
    return {
      'deviceId': deviceId,
      'environment': environment,
      'notificationsEnabled': notificationsEnabled,
      'warAttacksEnabled': warAttacks,
      'warStateEnabled': warState,
      'warRemindersEnabled': warReminders,
      'raidRemindersEnabled': raidReminders,
      'eventsEnabled': events,
      'announcementsEnabled': announcements,
      'monthlySupportEnabled': monthlySupport,
      'reminderTimings': reminderTimings,
      'raidReminderTimings': raidReminderTimings,
    };
  }

  Map<String, dynamic> toLocalJson() {
    return {
      ...toPutJson(deviceId: deviceId, environment: environment),
      'accounts': accounts
          .map(
            (account) => {
              'playerTag': account.playerTag,
              'source': account.source.name,
              'active': account.active,
            },
          )
          .toList(growable: false),
    };
  }

  bool enabled(NotificationCategory category) => switch (category) {
    NotificationCategory.warAttacks => warAttacks,
    NotificationCategory.warState => warState,
    NotificationCategory.warReminders => warReminders,
    NotificationCategory.raidReminders => raidReminders,
    NotificationCategory.events => events,
    NotificationCategory.announcements => announcements,
    NotificationCategory.monthlySupport => monthlySupport,
  };

  NotificationPreferences withCategory(
    NotificationCategory category,
    bool enabled,
  ) => switch (category) {
    NotificationCategory.warAttacks => copyWith(warAttacks: enabled),
    NotificationCategory.warState => copyWith(warState: enabled),
    NotificationCategory.warReminders => copyWith(warReminders: enabled),
    NotificationCategory.raidReminders => copyWith(raidReminders: enabled),
    NotificationCategory.events => copyWith(events: enabled),
    NotificationCategory.announcements => copyWith(announcements: enabled),
    NotificationCategory.monthlySupport => copyWith(monthlySupport: enabled),
  };

  NotificationPreferences copyWith({
    String? deviceId,
    String? environment,
    bool? notificationsEnabled,
    bool? warAttacks,
    bool? warState,
    bool? warReminders,
    bool? raidReminders,
    bool? events,
    bool? announcements,
    bool? monthlySupport,
    List<int>? reminderTimings,
    List<int>? raidReminderTimings,
    List<NotificationAccount>? accounts,
  }) {
    return NotificationPreferences(
      deviceId: deviceId ?? this.deviceId,
      environment: environment ?? this.environment,
      notificationsEnabled: notificationsEnabled ?? this.notificationsEnabled,
      warAttacks: warAttacks ?? this.warAttacks,
      warState: warState ?? this.warState,
      warReminders: warReminders ?? this.warReminders,
      raidReminders: raidReminders ?? this.raidReminders,
      events: events ?? this.events,
      announcements: announcements ?? this.announcements,
      monthlySupport: monthlySupport ?? this.monthlySupport,
      reminderTimings: reminderTimings ?? this.reminderTimings,
      raidReminderTimings: raidReminderTimings ?? this.raidReminderTimings,
      accounts: accounts ?? this.accounts,
    );
  }
}

File: expo/src/core/dto/notification-preferences.ts
export type NotificationCategory =
  | 'warAttacks'
  | 'warState'
  | 'warReminders'
  | 'raidReminders'
  | 'events'
  | 'announcements'
  | 'monthlySupport'
  | 'legendDefenses';

export interface NotificationAccount {
  readonly tag: string;
  readonly enabled: boolean;
}

export interface NotificationPreferences {
  readonly warAttacks: boolean;
  readonly warState: boolean;
  readonly warReminders: boolean;
  readonly raidReminders: boolean;
  readonly events: boolean;
  readonly announcements: boolean;
  readonly monthlySupport: boolean;
  readonly legendDefenses: boolean;
  readonly reminderTimings: readonly number[];
  readonly raidReminderTimings: readonly number[];
  readonly accounts: readonly NotificationAccount[];
}

export function createDefaultNotificationPreferences(): NotificationPreferences {
  return {
    warAttacks: false,
    warState: false,
    warReminders: false,
    raidReminders: false,
    events: false,
    announcements: false,
    monthlySupport: false,
    legendDefenses: false,
    reminderTimings: [],
    raidReminderTimings: [],
    accounts: [],
  };
}

export function isNotificationCategoryEnabled(
  preferences: NotificationPreferences,
  category: NotificationCategory,
): boolean {
  return preferences[category];
}

export function withNotificationCategory(
  preferences: NotificationPreferences,
  category: NotificationCategory,
  enabled: boolean,
): NotificationPreferences {
  return { ...preferences, [category]: enabled };
}

export function parseNotificationPreferences(value: unknown): NotificationPreferences {
  const json = expectRecord(value, 'notification preferences');
  return {
    warAttacks: expectBoolean(json.warAttacksEnabled, 'warAttacksEnabled'),
    warState: expectBoolean(json.warStateEnabled, 'warStateEnabled'),
    warReminders: expectBoolean(json.warRemindersEnabled, 'warRemindersEnabled'),
    raidReminders: expectBoolean(json.raidRemindersEnabled, 'raidRemindersEnabled'),
    events: expectBoolean(json.eventsEnabled, 'eventsEnabled'),
    announcements: expectBoolean(json.announcementsEnabled, 'announcementsEnabled'),
    monthlySupport: expectBoolean(json.monthlySupportEnabled, 'monthlySupportEnabled'),
    legendDefenses: expectBoolean(json.legendDefensesEnabled, 'legendDefensesEnabled'),
    reminderTimings: parseReminderTimings(json.reminderTimings, 2820, false, 'reminder timings'),
    raidReminderTimings: parseReminderTimings(
      json.raidReminderTimings,
      4320,
      true,
      'Raid Weekend reminder timings',
    ),
    accounts: expectArray(json.accounts, 'notification accounts').map(parseNotificationAccount),
  };
}

/** Reads the prior on-device snapshot only; network payloads remain strict RC19 shapes. */
export function parseLocalNotificationPreferences(value: unknown): NotificationPreferences {
  const json = expectRecord(value, 'local notification preferences');
  const accounts = Array.isArray(json.accounts)
    ? json.accounts.flatMap((value) => {
        if (!isRecord(value)) return [];
        if (typeof value.tag === 'string' && typeof value.enabled === 'boolean') return [value];
        if (
          value.source === 'verified' &&
          typeof value.playerTag === 'string' &&
          typeof value.active === 'boolean'
        ) {
          return [{ tag: value.playerTag, enabled: value.active }];
        }
        return [];
      })
    : [];
  return parseNotificationPreferences({
    ...json,
    raidRemindersEnabled: json.raidRemindersEnabled ?? false,
    raidReminderTimings: json.raidReminderTimings ?? [],
    legendDefensesEnabled: json.legendDefensesEnabled ?? false,
    accounts,
  });
}

export function serializeNotificationPreferencesForPut(preferences: NotificationPreferences) {
  return {
    warAttacksEnabled: preferences.warAttacks,
    warStateEnabled: preferences.warState,
    warRemindersEnabled: preferences.warReminders,
    raidRemindersEnabled: preferences.raidReminders,
    eventsEnabled: preferences.events,
    announcementsEnabled: preferences.announcements,
    monthlySupportEnabled: preferences.monthlySupport,
    legendDefensesEnabled: preferences.legendDefenses,
    reminderTimings: [...preferences.reminderTimings],
    raidReminderTimings: [...preferences.raidReminderTimings],
  };
}

export function serializeNotificationPreferencesForLocalStorage(
  preferences: NotificationPreferences,
): Record<string, unknown> {
  return {
    ...serializeNotificationPreferencesForPut(preferences),
    accounts: preferences.accounts.map((account) => ({ ...account })),
  };
}

export function parseNotificationAccount(value: unknown): NotificationAccount {
  const json = expectRecord(value, 'notification account');
  return {
    tag: expectString(json.tag, 'notification account tag', true),
    enabled: expectBoolean(json.enabled, 'notification account enabled'),
  };
}

function parseReminderTimings(
  value: unknown,
  maximum: number,
  requiresQuarterHour: boolean,
  label: string,
): readonly number[] {
  const timings = expectArray(value, label).map((item) => {
    if (typeof item !== 'number' || !Number.isInteger(item)) {
      throw new TypeError(`Invalid ${label}`);
    }
    return item;
  });
  if (
    timings.length > 3 ||
    timings.some(
      (timing) => timing < 1 || timing > maximum || (requiresQuarterHour && timing % 15 !== 0),
    ) ||
    new Set(timings).size !== timings.length
  ) {
    throw new TypeError(`Invalid ${label}`);
  }
  return timings;
}

function expectRecord(value: unknown, label: string): Record<string, unknown> {
  if (!isRecord(value)) throw new TypeError(`Invalid ${label}`);
  return value;
}

function expectString(value: unknown, label: string, nonEmpty = false): string {
  if (typeof value !== 'string' || (nonEmpty && value.length === 0)) {
    throw new TypeError(`Invalid ${label}`);
  }
  return value;
}

function expectBoolean(value: unknown, label: string): boolean {
  if (typeof value !== 'boolean') throw new TypeError(`Invalid ${label}`);
  return value;
}

function expectArray(value: unknown, label: string): readonly unknown[] {
  if (!Array.isArray(value)) throw new TypeError(`Invalid ${label}`);
  return value;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}
